src/appleseed/renderer/kernel/shading/shadingpoint.cpp

```diff
@@ -418,7 +418,10 @@ void ShadingPoint::refine_and_offset() const
         {
             assert(is_curve_primitive());
 
-            m_asm_inst_geo_normal = normalize(-asm_inst_ray.m_dir);
+            const Vector3d X = normalize(get_dpdu(0));
+            const Vector3d sn_curve = normalize(get_original_shading_normal());
```

Comment:
I believe that `get_original_shading_normal()` is already unit-length. Please double-check.

Take const ref.